Misc. tweaks to SourceStack

sjforeman opened the pull request:
This collects a few minor changes to `draco.analysis.sourcestack.SourceStack`:

- Allowing the use of input `FormedBeam` containers where the frequency axis is either increasing or decreasing. The convention for this is opposite in the `CHIME` telescope class and a generic `driftscan` telescope, so this change allows for stacking with non-`CHIME` telescopes.
- An option to only stack on sources within a specific frequency bin (useful for isolating the stacking signal within a narrow frequency range, as opposed to averaging across the entire band).
- Changing variables and comments to refer to sources instead of quasars/QSOs.

ssiegelx commented on draco/analysis/sourcestack.py:
```diff
@@ -77,101 +83,118 @@ def process(self, formed_beam):
-                self.log.info(
-                    f"Number of quasars stacked for pol {pstr}: {full_qcount}"
-                )
+                self.log.info(f"Number of quasars stacked for pol {pstr}: {full_count}")
```
```suggestion
                self.log.info(f"Number of sources stacked for pol {pstr}: {full_count}")
```
missed a quasar

ssiegelx commented on draco/analysis/sourcestack.py:
```diff
@@ -77,101 +83,118 @@ def process(self, formed_beam):
 
-            # Gather all ranks of qcount. Report number of quasars stacked
-            full_qcount = comm.reduce(qcount, op=MPI.SUM, root=0)
+            # Gather all ranks of count. Report number of quasars stacked
```
```suggestion
            # Gather all ranks of count. Report number of sources stacked.
```

ssiegelx commented on draco/analysis/sourcestack.py:
```diff
@@ -77,101 +83,118 @@ def process(self, formed_beam):
+        # This works because the frequency axis is not distributed between
+        # ranks.
+        if self.single_source_bin_index is not None:
+            restricted_chan_mask = (
```
I think this expression will fail for edge bins, in the sense that all sources that fall outside the range of frequencies contained in the formed beam will be included in the edge bin's `restricted_chan_mask`.  I'm not sure if this is an issue for your use case.

I think something like:
```
fs = formed_beam.index_map['freq'][self.single_source_bin_index]
restricted_chan_mask = np.abs(source_freq - fs['centre']) < (0.5 * fs['width'])
```
would handle that particular edge case properly.

ssiegelx reviewed (approved):
Hey Simon, this looks good, sorry it took me so long to review.

sjforeman commented:
Thanks for the review @ssiegelx , and sorry for the late follow-up! I implemented your suggestions and rebased to clean up the commit history. I'll merge this once all the CI checks are done
